plugins/woocommerce/src/Templating/TemplatingRestController.php

```diff
@@ -0,0 +1,419 @@
+	 *
+	 * @var array
+	 */
+	private array $authentication_errors_by_method;
+
+	/**
+	 * Class constructor.
+	 */
+	public function __construct() {
+		$this->authentication_errors_by_method = array(
+			'GET'    => array(
+				'code'    => 'woocommerce_rest_cannot_view',
+				'message' => __( 'Sorry, you cannot view resources.', 'woocommerce' ),
+			),
+			'POST'   => array(
+				'code'    => 'woocommerce_rest_cannot_create',
+				'message' => __( 'Sorry, you cannot create resources.', 'woocommerce' ),
+			),
+			'DELETE' => array(
+				'code'    => 'woocommerce_rest_cannot_delete',
+				'message' => __( 'Sorry, you cannot delete resources.', 'woocommerce' ),
+			),
+		);
+
+		self::add_filter( 'woocommerce_rest_api_get_rest_namespaces', array( $this, 'handle_woocommerce_rest_api_get_rest_namespaces' ) );
+
+		self::add_action( 'init', array( $this, 'handle_init' ), 0 );
+		self::add_filter( 'query_vars', array( $this, 'handle_query_vars' ), 0 );
+		self::add_action( 'parse_request', array( $this, 'handle_parse_request' ), 0 );
+	}
+
+	/**
+	 * Class initialization, to be executed when the class is resolved by the container.
+	 *
+	 * @param TemplatingEngine $templating_engine The instance of TemplatingEngine to use.
+	 * @internal
+	 */
+	final public function init( TemplatingEngine $templating_engine ) {
+		$this->templating_engine = $templating_engine;
+	}
+
+	/**
+	 * Handle the woocommerce_rest_api_get_rest_namespaces filter
+	 * to add ourselves to the list of REST API controllers registered by WooCommerce.
+	 *
+	 * @param array $namespaces The original list of WooCommerce REST API namespaces/controllers.
+	 * @return array The updated list of WooCommerce REST API namespaces/controllers.
+	 */
+	private function handle_woocommerce_rest_api_get_rest_namespaces( array $namespaces ): array {
+		$namespaces['wc/v3']['templates'] = self::class;
+		return $namespaces;
+	}
+
+	/**
+	 * Register the JSON REST API endpoints handled by this controller.
+	 */
+	public function register_routes() {
+		self::mark_method_as_accessible( 'run' );
+
+		register_rest_route(
+			$this->route_namespace,
+			'/templates/(?P<id>[\d]+)',
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => fn( $request ) => $this->run( 'get_template', $request ),
+				'permission_callback' => fn( $request ) => $this->check_permission( $request, 'read_rendered_template_info' ),
+				'args'                => $this->get_args_for_get_or_delete_template(),
+			)
+		);
+
+		register_rest_route(
+			$this->route_namespace,
+			'/templates/(?P<id>[\d]+)',
+			array(
+				'methods'             => WP_REST_Server::DELETABLE,
+				'callback'            => fn( $request ) => $this->run( 'delete_template', $request ),
+				'permission_callback' => fn( $request ) => $this->check_permission( $request, 'delete_rendered_template' ),
+				'args'                => $this->get_args_for_get_or_delete_template(),
+			)
+		);
```

Comment:
Since these have the same namespace and route, they can be combined into one `register_rest_route` call, like this:

```suggestion
	register_rest_route(
		$this->route_namespace,
		'/templates/(?P<id>[\d]+)',
		array(
			array(
				'methods'             => WP_REST_Server::READABLE,
				'callback'            => fn( $request ) => $this->run( 'get_template', $request ),
				'permission_callback' => fn( $request ) => $this->check_permission( $request, 'read_rendered_template_info' ),
				'args'                => $this->get_args_for_get_or_delete_template(),
			),
			array(
				'methods'             => WP_REST_Server::DELETABLE,
				'callback'            => fn( $request ) => $this->run( 'delete_template', $request ),
				'permission_callback' => fn( $request ) => $this->check_permission( $request, 'delete_rendered_template' ),
				'args'                => $this->get_args_for_get_or_delete_template(),
			),
		)
	);
```

It would also be good to have a `schema` property for the route, like we do for other endpoints, e.g. [Orders](https://github.com/woocommerce/woocommerce/blob/1a48f47d322b06ebd3492ecb87f594113d1ac13b/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-orders-v2-controller.php#L82-L82).